cfme/common/provider.py

```diff
@@ -361,20 +365,61 @@ def _fill_rsa_endpoint_dicts(self, provider_attributes, connection_configs):
         Helper method for ``self.create_rest``
         """
         if "rsa_keypair" not in self.endpoints:
-            return
+            if "rsa_keypair" not in self.data.endpoints:
```

Comment:
:+1: :100:

Reply:
Why is this additional check against data done here? This should just be processing the endpoints that were set when the provider object was instantiated via the BaseProvider attribute.

This is not the right way to approach handling filling this endpoint value.

`BaseProvider.endpoints` is a constructor attribute, and is set when the provider object is instantiated. This is generally via `cfme.utils.providers.get_crud` method, which will use the individual provider class `from_config` method. In the case of openstack, that's here:

https://github.com/ManageIQ/integration_tests/blob/master/cfme/cloud/provider/openstack.py#L86

This is obviously only creating a single entry in the endpoints dictionary, using the RHOSEndpoint class that's actually defined in the openstack_infra module at the next link. Note OpenstackInfraProvider has its own `from_config` method, which handles all three types of endpoints for openstack. I assume this change is being made to support the rsa endpoint for the IMS provider, which is _not_ OpenstackInfra.

https://github.com/ManageIQ/integration_tests/blob/master/cfme/infrastructure/provider/openstack_infra.py#L17

This needs to be addressed by updating the `OpenstackProvider.from_config()` method to actually process any endpoints other than `default` for openstack.  There (likely) won't be any changes to the method here, as its already processing the endpoint information in exactly the same way, you have just added references to the yaml data directly instead of the provider instance.